spec/requests/application_controller/filter_spec.rb

```diff
@@ -0,0 +1,120 @@
+    it("key:eq y-m-dTh:m:s single with 'eq' key")    { expect_success("filter[completed_at][eq]=2019-03-06T16:17:00", task_3) }
+    it("key:eq y-m-dTh:m single with 'eq' key")      { expect_success("filter[completed_at][eq]=2019-03-06T16:17", task_3) }
+
+    it("key:eq y-m-dTh:m:s array without 'eq' key") { expect_success("filter[completed_at][]=2019-03-06T16:17:00&filter[completed_at][]=2019-03-06T16:15:00", task_2, task_3) }
+    it("key:eq y-m-dTh:m array without 'eq' key")   { expect_success("filter[completed_at][]=2019-03-06T16:17&filter[completed_at][]=2019-03-06T16:15", task_2, task_3) }
+    it("key:eq y-m-dTh:m:s array with 'eq' key")    { expect_success("filter[completed_at][eq][]=2019-03-06T16:17:00&filter[completed_at][eq][]=2019-03-06T16:15:00", task_2, task_3) }
+    it("key:eq y-m-dTh:m array with 'eq' key")      { expect_success("filter[completed_at][eq][]=2019-03-06T16:17&filter[completed_at][eq][]=2019-03-06T16:15", task_2, task_3) }
+
+    it("key:gt y-m-dTh:m:s") { expect_success("filter[completed_at][gt]=2019-03-06T16:16:05", task_3, task_4) }
+    it("key:gt y-m-dTh:m")   { expect_success("filter[completed_at][gt]=2019-03-06T16:16", task_3, task_4) }
+    it("key:gt y-m-d")       { expect_success("filter[completed_at][gt]=2019-03-06", task_2, task_3, task_4) }
+
+    it("key:lt y-m-dTh:m:s") { expect_success("filter[completed_at][lt]=2019-03-06T16:17:05", task_2, task_3) }
+    it("key:lt y-m-dTh:m")   { expect_success("filter[completed_at][lt]=2019-03-06T16:17", task_2) }
+    it("key:lt y-m-d")       { expect_success("filter[completed_at][lt]=2019-03-06") }
+
+    it("key:nil")            { expect_success("filter[completed_at][nil]", task_1) }
+    it("key:not_nil")        { expect_success("filter[completed_at][not_nil]", task_2, task_3, task_4) }
+  end
+
+  context "error cases" do
+    it("empty filter")      { expect_failure("filter", "found unpermitted parameter: :filter") }
+    it("unknown attribute") { expect_failure("filter[xxx]", "found unpermitted parameter: xxx") }
+
+    context "unsupported comparator" do
+      it("on an integer")  { expect_failure("filter[id][xxx]=4", "unsupported integer comparator: xxx") }
+      it("on a string")    { expect_failure("filter[name][xxx]=4", "unsupported string comparator: xxx") }
+      it("on a timestamp") { expect_failure("filter[created_at][xxx]=4", "unsupported timestamp comparator: xxx") }
+    end
+
+    it "unsupported attribute type" do
+      source_type = SourceType.create!(:name => "a", :vendor => "a", :product_name => "a")
+      source = Source.create!(:source_type => source_type, :tenant => tenant, :name => "a")
+      Vm.create!(:source => source, :tenant => tenant, :source_ref => "a")
+
+      get("/api/v0.1/vms?filter[mac_addresses]=a")
+
+      expect(response.status).to eq(400)
+      expect(response.parsed_body["errors"]).to eq([{"detail"=>"unsupported attribute type for: mac_addresses", "status"=>400}])
```

Comment:
We might want a spec for multiple bad keys, or perhaps a mix of good and bad keys.

Reply:
I agree, that would be great.